### PR TITLE
DrivAerML: 5-epoch warmup + gc=1.0 compound (stabilized warmup)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

5-epoch warmup alone (#3069) achieved 5.823% best before diverging at ep523 — the warmup direction was not wrong, just unstable. The warmup hypothesis is: gradual LR ramp allows the optimizer to establish stable update directions before hitting full lr=5e-4. The gap: warmup without gc still hits catastrophic LR-peak explosions.

This compounds warmup + gc=1.0 to provide the missing stability floor. gc=1.0 (hard gradient ceiling) + 5-epoch warmup (gradual start) together should enable stable training past the ep150+ region where warmup-only runs diverged.

## Instructions

**Run: 5-epoch linear warmup + gc=1.0 + T_max=30**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 1.0 \
  --warmup-epochs 5 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group dm-warmup-gc-compound
```

Note: best gc result on DM without warmup was 4.346% (gc=2.0, PR #2886) which still didn't beat baseline 3.997%. The warmup may push gc below baseline if the combination is synergistic. Also run without warmup for comparison:

**Control: gc=1.0 only (no warmup)**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 1.0 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group dm-warmup-gc-compound
```

## Baseline
- **val_primary/surface_rel_l2_pct:** 3.997% at epoch 467 (no gc, no warmup)
- **Best gc result:** 4.346% gc=2.0 (#2886)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT chopper:

## Results

### Treatment: 5-epoch warmup + gc=1.0 + T_max=30

| Metric | Baseline | This Run | Delta |
|--------|----------|----------|-------|
| val_primary/surface_rel_l2_pct | 3.997% | 9.505% (ep48) | +5.508% ❌ |
| test_primary/surface_rel_l2_pct | — | 10.385% (ep48) | — |

- **Command:** `CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=360 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 1.0 --warmup-epochs 5 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --agent chopper --wandb-name "chopper/dm-warmup5-gc1.0-tmax30" --wandb-group dm-warmup-gc-compound`
- **W&B:** `f4jfokoj`
- **Peak memory:** ~H100 97GB available, run completed without OOM

### Control: gc=1.0 only (no warmup)

| Metric | Baseline | This Run | Delta |
|--------|----------|----------|-------|
| val_primary/surface_rel_l2_pct | 3.997% | 9.064% (ep89) | +5.067% ❌ |
| test_primary/surface_rel_l2_pct | — | 9.533% (ep89) | — |

- **Command:** `CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=360 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 1.0 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --agent chopper --wandb-name "chopper/dm-gc1.0-only-tmax30" --wandb-group dm-warmup-gc-compound`
- **W&B:** `kkkmsuxc`
- **Peak memory:** ~H100 97GB available, run completed without OOM

### What happened

**Both runs are clear negatives.** gc=1.0 is too aggressive for DrivAerML — neither warmup nor no-warmup came close to the 3.997% baseline.

**Key observations:**

1. **gc=1.0 clips every batch** — in early epochs, 394/394 batches were clipped, meaning the gradient ceiling is cutting the natural gradient magnitude by >100x. This severely restricts learning capacity.

2. **Both runs diverged catastrophically:**
   - Treatment (warmup + gc): Best at ep48 (9.5%), then jumped from 22% to 66% at ep50-51. The warmup ended at step 1970 (ep5), and the model made progress ep5→48 but hit a cosine LR peak that triggered an unrecoverable explosion.
   - Control (gc only): Best at ep89 (9.1%), then jumped from 9.1% to 42% at ep91. Same pattern — a cosine LR peak at full lr=5e-4 destabilized the model.

3. **Warmup made things slightly worse**, not better: treatment diverged at ep50 vs control at ep89. The warmup may have prevented the model from learning robust representations during the constrained early phase, making it more fragile when the cosine schedule hit high LR.

4. **The divergence pattern is identical to #3069** (warmup-only): a sudden explosion at a cosine LR peak that the model never recovers from. gc=1.0 doesn't prevent this — it just delays it.

### Analysis

gc=1.0 fundamentally doesn't work for DrivAerML at lr=5e-4. The natural gradient norms are much larger than 1.0, so the clipping is too restrictive — the model underfits during stable phases and then explodes when the cosine schedule pushes LR up. The best prior gc result (gc=2.0, 4.346%) was better precisely because it was less restrictive, but even that didn't beat baseline. The compound warmup+gc hypothesis is not synergistic — the stability floor from gc=1.0 is too low to learn well, and warmup doesn't compensate.

### Suggested follow-ups

- **Higher gc thresholds with warmup**: gc=2.0 or gc=3.0 + warmup might be worth testing — these clip less aggressively and may actually benefit from warmup's gradual start.
- **Adaptive gradient clipping** (e.g. AGC): clips based on parameter norm, not absolute norm. This would avoid the "one-size-fits-all" problem with gc=1.0.
- **Lower LR with gc=1.0**: If gc=1.0 is desired, lr=1e-4 or lr=2e-4 might avoid the cosine peak explosions since the peaks would be much lower.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3137 because ADVISOR: gc=1.0 without EMA confirmed dead. warmup+gc=1.0: 9.505% ep48, diverged ep50 (W&B: f4jfokoj). gc=1.0 only: 9.064% ep89, diverged ep91 (W&B: kkkmsuxc). gc=1.0 clips 100% of batches — underfitting during stable phases then explosion at cosine peaks. 2.4x worse than baseline. The entire gc=1.0 family without EMA is invalidated. Reassigning to bold noam-feature experiment..
